### PR TITLE
[Enhancement] New Option for Help, Initial Module, and Tests for the Source Code Documentation

### PR DESCRIPTION
**Pull request by kevinmatthes**
Hello, @wert007,

This Pull Request introduces the following new features:

* a "help" option, as many CLIs have them,
* an initial module in order to enhance the maintainability of the application, and
* some initial source code documentation strings ("docstrings") for `cargo doc`.

This Pull Request is also intended to discuss the source code style in order to agree on some initial contribution guidelines.

### CODE REVIEW

**Comment by wert007** [comment]
Okay, before I start dissecting your pull request: Please run `cargo fmt` to auto format the code.

Secondly run `cargo clippy` which will lint your code to be up to rust standards. And follow the recommendations. 😄

**Comment by kevinmatthes** [comment]
Okay, I applied your suggestions.  Please note that the linter still complains about 6 regions which I do not edited with none of my commits.

I order to do not modify the behaviour of the application unintendedly, I would prefer to leave them as they are, at the moment.

**Comment by kevinmatthes** [comment]
Mostly, the complaints concern the closures of `.map_err`.  Once, the member `merge: Option <String>` is considered obsolete.

File: src/usage.rs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+//!
+//! * the application name is hard coded,
+//! * the tool tip is hard coded, and
+//! * the expected order of arguments in the usage template is hard coded.
+
+use getopts::Options;
+
+/// A small in-app documentation.
+///
+/// * `options` - The options the render the tool tips for.
```
Comment (wert007) [inline src/usage.rs:14]: Hmm, it seems rust does not really have a standard for documenting function parameters. (See [here](https://github.com/rust-lang/rust/issues/57525) for progress on that matter)

Since the parameter is already covered by the paragraph below I would remove this. 😅

**Review by wert007 — CHANGES_REQUESTED** [review]
Added some feedback here 😄

**Comment by wert007** [comment]
> Okay, I applied your suggestions. Please note that the linter still complains about 6 regions which I do not edited with none of my commits.
> 
> I order to do not modify the behaviour of the application unintendedly, I would prefer to leave them as they are, at the moment.

Yeah I saw that. I will fix these after the pull request merged I think 😅

File: src/main.rs [hunk 2]
```diff
@@ -1,9 +1,13 @@
+use crate::usage::usage;
```
Comment (wert007) [inline src/main.rs:1]: I think this could be simplified to
```rust
use usage::usage;
```

File: src/main.rs [hunk 3]
```diff
@@ -79,14 +83,14 @@ fn main() -> Result<(), Box<dyn Error>> {
...
         Ok(it) => it,
         Err(err) => {
-            usage("commit-analzer", opts);
+            usage(opts);
             return Err(err.into());
         }
     };
-    if matches.free.len() < 2 {
-        usage(&matches.free[0], opts);
+    if matches.opt_present("h") || matches.free.len() < 0x2 {
```
[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/main.rs:90]: In C, there is a feature named "Numeral Annotation".  It is considered good style to append the corresponding letters to a number in order to indicate the intended range.  For instance, `123ULL` will then be assumed to be of the type `unsigned long long int` and not just a simple `int`.

The only two numeric primitives which do not have such an annotation are `int` and `double`.  `double` numbers also accept the `int` representation but may behave other than expected then.  In order to distinct the representation of an `int` from the representation of a `double`, I prefer the hexadecimal representation for integers.  Hexadecimal number will always be parsed to `int`s, so the meaning of `0x2A` (42) is always clear to both the compiler and the software engineers reading the code:  `0x2A` is intended to be a simple `int`.

The hexadecimal representation is also wonderful to abbreviate numbers, especially those which are powers to two.  Often, C / C++ headers specify a set of magic numbers using hexadecimal numbers, for instance OpenGL.

I used the hexadecimal representation here in Rust, as well, since I always use it to indicate a number being intended to be compiled as an integer.  More precisely, I specify integers numbers usually as hexadecimals since, due to the C / C++ coding, it became normal to me.

If you dislike it, I can change the representation back to decimal.

File: src/main.rs [hunk 4]
```diff
@@ -79,14 +83,14 @@ fn main() -> Result<(), Box<dyn Error>> {
...
+        usage(opts);
         return Ok(());
-    }
+    };
```
[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/main.rs:93]: Also a habit from C / C++.  I always set all semicolons where they are allowed and not linted away.

Many C / C++ software engineers probably know that:  they specify a new type (both `struct` and `class`) together with all associated methods / member functions / utility functions, but there are some smaller bugs due to which the compiler denies the compilation.  Okay, one fixes them one after another but still, the new type does not compile:  one forgot to end the `class` / `struct` definition with a semicolon.

If you always set just the required amount of semicolons, you need to learn by heart when you need one.  The other way around, when you always set all semicolons, even those which are obsolete, you need to learn where to not set one.

It turns out that learning the exceptions is more effective than learning all places you explicitly need one because there are usually more required places than places where you should omit the semicolon.

If you dislike it, I can remove it.

File: src/usage.rs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+///
+/// This function will write a brief usage information, including a short
+/// introduction to the meaning of the configured options, to `stdout`.
+pub fn usage(options: &Options) {
+    println!(
+        "Usage:  commit-analyzer <FILE> [OPTIONS]\n{}",
```
[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/usage.rs:20]: I began the set of changes with the fix of a typo:  `commit-analzer`.  One had forgotten the letter `y`.  In order to prevent similar typos in the future, I looked for an appropriate solution.

Making the name of the application a constant such that you only need to care for the correctness of your constant would help.  But then, you could also hard code the name directly into the corresponding function since it cannot be reused in another project due to having a path to a file as a mandatory argument.

Making it completely flexible and, hence, reusable would have meant a parameter overkill.  You have needed to pass

* the application's name,
* its tool tip,
* its options, and
* the usage line.

Hard coding the app name seemed to be most effective.

File: src/usage.rs [hunk 6]
```diff
@@ -0,0 +1,23 @@
+//! The module for the usage advice.
```
[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/usage.rs:1]: It was intended to test the capabilities of the module system and the docstrings at once without uglifying the main source file.

It could be renamed to `utilities.rs` as a storage for utility functions.

File: src/usage.rs (hunk 1 above)
Comment (kevinmatthes) [reply src/usage.rs:14]: I oriented on the documentation of `getopts :: Options` and the way a good Doxygen documentation is built.

File: src/main.rs [hunk 7]
```diff
@@ -1,9 +1,13 @@
...
 use std::{collections::HashMap, error::Error, io::Write, num::ParseIntError, ops::AddAssign};
 
+mod usage;
+
 fn main() -> Result<(), Box<dyn Error>> {
     let mut opts = getopts::Options::new();
     let opts = opts
-        .optflag("q", "quiet", "Hides the output of commit messages")
+        .optflag("q", "quiet", "Hides the output of commit messages.")
+        .optflag("h", "help", "Show this help and exit.")
```
[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/main.rs:10]: Yes, this also happened to me.  I expected the CLI to have a help option and was confused that it was considered a failure.  So, I added this feature since I considered it a meaningful change.

File: src/main.rs (hunk 2 above)
Comment (kevinmatthes) [reply src/main.rs:1]: I would rather leave it in.  The plan is to also outsource the types you introduced to dedicated modules.  When importing them, the `crate ::` part will indicate to human readers that these modules belong to this project.

File: src/usage.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/usage.rs:20]: I removed one `println` because it is a function call which can be omitted by just combining the two texts.  That way, a slight performance boost is achieved.  (Obviously, it is cheaper in terms of performance to call only one function instead of two.)

I decided to still use `println` since `cargo` uses the same formatting for its option documentation; see `cargo --help` for details or, more precisely, the blank line before between its output and your prompt.

File: src/main.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (wert007) [reply src/main.rs:90]: In rust this is not a problem, since
```rust
let d: f64 = 42;
```
would not compile. You would have to write either `42.0` or `42f64` for that.

File: src/main.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (wert007) [reply src/main.rs:93]: Hmm, for one, you can always insert almost everywhere semicolon. So would you write also such c code?
```c
#include<stdlib.h>;
;
void foo(){};
```

And secondly, this is kind of part of learning a programming language. Knowing how it works and why it works, this would also includes semicolons. 😅

File: src/usage.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (wert007) [reply src/usage.rs:20]: Yeah, lke I said, I'm fine with this change.

> Also you removed a line feed, since `println` will add a `\n` at the end of your format string.

Please fix that. While this might seem petty, I think pull request should be similarly atomic like commits. So accidental changes should be reverted here.

File: src/usage.rs (hunk 1 above)
Comment (wert007) [reply src/usage.rs:14]: The very good [documentation of the std lib](docs.rs/std) uses itself little to non explicit parameter documentation. I think that style would be more important than doxygen. 😅

File: src/main.rs (hunk 2 above)
Comment (wert007) [reply src/main.rs:1]: That would be a good candidate for creating a separate module. But I think the signal, that there are files by that name in the same directory should be signal enough to human readers. 😅

File: src/usage.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (wert007) [reply src/usage.rs:1]: Right now there not really any utility functions, so I would rather not. I think what you wrote down below with putting the types into their own module makes sense. But I would maybe recommend to you, to first implement the support for the `stdin` input, since this will likely not require a seperate module. After that you have a better overview how big the project actually became, and in what way modules can help or maybe be a hindrance to readabiltiy. 😄

File: src/usage.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (wert007) [reply src/usage.rs:20]: > I removed one `println` because it is a function call which can be omitted by just combining the two texts. That way, a slight performance boost is achieved. (Obviously, it is cheaper in terms of performance to call only one function instead of two.)

Few things. Firstly, println is not a function, but a macro.
Secondly, I actually checked the assembler output, because I thought the llvm would be able to optimize this into the same code. But I think it fails to do that, since println also flushes its writer and this may lead to difference to other programs reading from its output. I think. But in the end the difference is a single `call` instruction, which is not perceivable in any way, since any natural fluctuation is worse then that. Also this is a **very** cold code path, so that performance boosts here are not important at all.

But I'm fine with keeping it as one `println!`. But performance would not be an argument in favor of that. 😅

File: src/main.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/main.rs:93]: First, the shown code is invalid according to `-std=c99`.  "ISO C denies semicolons outside of functions" would be the answer of the compiler to that, at least GCC thinks that way.  But basically, yes, this looks like good C code to me.

Second, did you read "Crafting Interpreters" by Robert Nystrom?  (https://github.com/munificent/craftinginterpreters)  He writes that similarities between coding languages, especially when designing a new one, help to reduce the learning effort.  And to some point, Rust looks quite like C++, so basically, knowing C++ and how to apply its semicolon rules should suffice to deduce the basics of how Rust should handle instruction delimitation.

[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/main.rs:93]: Due to this discussion on that single semicolon, I assume I should rather remove it, right?

File: src/usage.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/usage.rs]: I just fixed it.

File: src/usage.rs (hunk 1 above)
Comment (kevinmatthes) [reply src/usage.rs]: Done.

File: src/main.rs (hunk 2 above)
Comment (kevinmatthes) [reply src/main.rs]: Done.

By the way, I sorted the import lines by alphabet, this is, moving the `use usage::usage;` under the import of the `std` crates.  In my opinion, that way, it is easier to keep track of which modules are imported.

File: src/main.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/main.rs]: Done.

File: src/main.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (wert007) [reply src/main.rs]: > Due to this discussion on that single semicolon, I assume I should rather remove it, right?

Yeah 👍

[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/main.rs]: Done.

I wonder why `rustfmt` alias `cargo fmt` does not lint such things away.  Maybe a candidate for a Feature Request.

[the remark this answers is not in the document]
Comment (wert007) [reply src/main.rs]: > Second, did you read "Crafting Interpreters" by Robert Nystrom? (https://github.com/munificent/craftinginterpreters) He writes that similarities between coding languages, especially when designing a new one, help to reduce the learning effort. And to some point, Rust looks quite like C++, so basically, knowing C++ and how to apply its semicolon rules should suffice to deduce the basics of how Rust should handle instruction delimitation.

I did indeed. But this should not stop one from learning the other language completely. But I'm not sure, if I agree completely, e.g. enums are quite different between c++ and rust, not only on the syntactic level, but also there. I think it should make reading a foreign programming language easier, but I don't think it should affect how you write code in the long term. 😅

File: src/main.rs (hunk 2 above)
Comment (wert007) [reply src/main.rs]: All right. 😄 

You might check if `rust fmt` changes the order. I think the `rust-analyser` extension for vs code does in its default settings.

File: src/usage.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/usage.rs:1]: At the moment, there are four standalone functions in the project:

* `main`
* `matches_filter`
* `parse_commit`
* `usage

From those four, I would just consider `main` and `parse_commit` actual business logic and the other two, `matches_filter` and `usage`, merely syntactic sugar.  In C, I would `inline` `matches_filter` (because a macro might suffer from side effects which could cause different results thereby) and turn `usage` into a macro (or `inline` it, as well).

I agree with you that an own module for a syntactic sugar function might be too much.  But my suggestion is to create one common module for such simple functions which only exist to abbreviate something.

In summary, I would like to suggest to move `matches_filter` into the `usage` module, as well, and to rename that module appropriately.

File: src/main.rs (hunk 2 above)
Comment (kevinmatthes) [reply src/main.rs]: Would it be an idea to add something like a pre-commit check or GitHub Action to this repository to check for such things and correct them automatically, at option?

File: src/main.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/main.rs:10]: By the way, how about adding an option overview to the README?

File: src/usage.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (wert007) [reply src/usage.rs:1]: Hmm, we could make `matches_filter` a member method in Commit. So basically something like this:

```rust
impl Commit {
    pub fn matches_filter(&self, filter: &Filter) -> bool {
        // Old code here
    }
}
```

Because I would create modules rather topic based. So that, e.g., in a module there are all functions regarding the frontend, or parsing or something like this. This makes it easier to find relevant functions, types, and also modules, for the current source file. Meanwhile in your utilities functions approach there would be a lot of functions, that would be called from something completely else, and functions, that are written close by do not have to have a correlation with each other in any way. 😅 

You get, where I'm getting at?

[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/usage.rs:1]: I see your point.  But for this Pull Request, shall I revert the module?  If so, what about its docstrings?

[the remark this answers is not in the document]
Comment (wert007) [reply src/usage.rs:1]: Hmm, I think, a lot of the stuff written in it, is to be expected by anyone already. Maybe you can combine it with the comment for the function, but if so, please shorten it by a lot. 😄

[the remark this answers is not in the document]
Comment (wert007) [reply src/usage.rs:1]: Yes, I think, that is the right course of action 👍

[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/usage.rs]: Done.

**Comment by wert007** [comment]
Thanks, I merged it. And will now look at the unused merge field. 😄

File: src/main.rs (hunk 2 above)
Comment (wert007) [reply src/main.rs]: Sure, I just never worked with these things. 😅

File: src/main.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (wert007) [reply src/main.rs:8]: Sure, why not 😄
